[quant][fx][pt2e][be] Store node_name_to_target_dtype to node.meta["target_dtype_info"]

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92641
* #92589
* #92575
* __->__ #92574

Summary:
This is in preparation for quantize_pt2e API where we allow programability for users to set how
they want to quantize their model

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][fx][pt2e][be] Store node_name_to_target_dtype to node.meta["target_dtype_info"]

Summary:
This is in preparation for quantize_pt2e API where we allow programability for users to set how
they want to quantize their model

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][pt2e][be] Store node_name_to_target_dtype to node.meta["target_dtype_info"]"


Summary:
This is in preparation for quantize_pt2e API where we allow programability for users to set how
they want to quantize their model

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][pt2e][be] Store node_name_to_target_dtype to node.meta["target_dtype_info"]"


Summary:
This is in preparation for quantize_pt2e API where we allow programability for users to set how
they want to quantize their model

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][pt2e][be] Store node_name_to_target_dtype to node.meta["target_dtype_info"]"


Summary:
This is in preparation for quantize_pt2e API where we allow programability for users to set how
they want to quantize their model

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1098,30 +1072,27 @@ def insert_observers_for_model(
-    #       'bias_dtype': torch.float,
-    #       'output_activation_dtype': torch.quint8,
-    #     }
+    #   conv2d_node.meta["target_dtype_info"] = {
```
nit: should this be `conv2d` here and also `weight` above (L1082)? or maybe change `conv2d` to `conv2d_node` in above line.

jerryzh168 replied:
sure

jcaip commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1137,7 +1108,8 @@ def insert_observers_for_model(
         input_quantized_idxs: List[int] = prepare_custom_config.input_quantized_indexes
         output_quantized_idxs: List[int] = prepare_custom_config.output_quantized_indexes
-        node_name_to_target_dtype_info[node.name] = _get_target_activation_dtype_for_node(
+        # Dict[str, Optional[Tuple[Union[torch.dtype, type], bool]]]
```
can you just put this as mypy type instead of a comment?

```
node.meta["target_dtype_info"] : Dict[str, Optional[Tuple[Union[torch.dtype, type], bool]]] = ...
```

jerryzh168 replied:
yeah I started with this, but it doesn't work

jcaip replied:
maybe try splitting into two lines?

```
node_target_dtype_info : Dict[str, Optional[Tuple[Union[torch.dtype, type], bool]]] = ...
node.meta["target_dtype_info"] = node_target_dtype_info
```

jcaip replied:
Or I think its fine with no comment too, just think if the return type changes the comment will be confusing

jerryzh168 replied:
OK, will try

jerryzh168 replied:
I think this is working, thanks

jcaip commented on torch/ao/quantization/quantize_fx.py:
```diff
@@ -99,6 +111,8 @@ def _prepare_fx(
             "in a future version. Please pass in a PrepareCustomConfig instead.")
         prepare_custom_config = PrepareCustomConfig.from_dict(prepare_custom_config)
 
+    _attach_meta_to_node_if_not_exist(model)
```
do you need this here?

jerryzh168 replied:
oh this should be after tracing, will move

jcaip replied:
I think since you call in `_fuse_fx` you don't need to call it twice.

jerryzh168 replied:
oh, I should put that in `_fuse_fx` I think, will move that one as well

jerryzh168 committed:
Update on "[quant][fx][pt2e][be] Store node_name_to_target_dtype to node.meta["target_dtype_info"]"


Summary:
This is in preparation for quantize_pt2e API where we allow programability for users to set how
they want to quantize their model

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][pt2e][be] Store node_name_to_target_dtype to node.meta["target_dtype_info"]"


Summary:
This is in preparation for quantize_pt2e API where we allow programability for users to set how
they want to quantize their model

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip reviewed (approved):
lgtm

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge -f

jerryzh168 commented:
@pytorchbot merge -f "not sure why there is a canceled check, but rest are succeeding"
